src/emulator/io/src/io.cpp

```diff
@@ -390,333 +288,236 @@ SceOff seek_file(SceUID fd, SceOff offset, int whence, IOState &io, const char *
-            statp->st_mode |= SCE_S_IFDIR;
-        } else {
-            statp->st_mode |= SCE_S_IFREG;
-        }
+    WIN32_FIND_DATAW find_data;
+    const auto handle = FindFirstFileW(file_path.generic_path().wstring().c_str(), &find_data);
+    if (handle == INVALID_HANDLE_VALUE)
+        return IO_ERROR(SCE_ERROR_ERRNO_EMFILE);
 
-        statp->st_size = (std::uint64_t)find_data.nFileSizeHigh << 32 | (std::uint64_t)find_data.nFileSizeLow;
+    FindClose(handle);
 
-        last_access_time_ticks = convert_filetime(find_data.ftLastAccessTime);
-        last_modification_time_ticks = convert_filetime(find_data.ftLastWriteTime);
-        creation_time_ticks = convert_filetime(find_data.ftCreationTime);
+    last_access_time_ticks = convert_filetime(find_data.ftLastAccessTime);
+    creation_time_ticks = convert_filetime(find_data.ftCreationTime);
 #else
-        struct stat sb;
-        if (stat(file_path.c_str(), &sb) < 0) {
-            return IO_ERROR_UNK();
-        }
-
-        if (S_ISREG(sb.st_mode)) {
-            statp->st_mode |= SCE_S_IFREG;
-        } else if (S_ISDIR(sb.st_mode)) {
-            statp->st_mode |= SCE_S_IFDIR;
-        }
-
-        statp->st_size = sb.st_size;
+    struct stat sb;
+    if (stat(file_path.generic_path().string().c_str(), &sb) < 0)
+        return IO_ERROR_UNK();
 
-        last_access_time_ticks = (uint64_t)sb.st_atime * VITA_CLOCKS_PER_SEC;
-        last_modification_time_ticks = (uint64_t)sb.st_mtime * VITA_CLOCKS_PER_SEC;
-        creation_time_ticks = (uint64_t)sb.st_ctime * VITA_CLOCKS_PER_SEC;
+    last_access_time_ticks = (uint64_t)sb.st_atime * VITA_CLOCKS_PER_SEC;
+    creation_time_ticks = (uint64_t)sb.st_ctime * VITA_CLOCKS_PER_SEC;
 
 #undef st_atime
 #undef st_mtime
 #undef st_ctime
 #endif
-        break;
-    }
-    default: {
-        LOG_ERROR("Unimplemented device {} used", file);
-        return IO_ERROR_UNK();
-    }
+
+    statp->st_mode = SCE_S_IRUSR | SCE_S_IRGRP | SCE_S_IROTH | SCE_S_IXUSR | SCE_S_IXGRP | SCE_S_IXOTH;
+
+    const auto last_modification_time_ticks = fs::last_write_time(file_path);
+
+    if (fs::is_regular_file(file_path)) {
+        statp->st_size = fs::file_size(file_path);
+
+        if (fd != invalid_fd)
+            statp->st_attr = SCE_SO_IFREG;
     }
+    if (fs::is_directory(file_path) && fd != invalid_fd)
+        statp->st_attr = SCE_SO_IFDIR;
+
     __RtcTicksToPspTime(&statp->st_atime, last_access_time_ticks);
     __RtcTicksToPspTime(&statp->st_mtime, last_modification_time_ticks);
     __RtcTicksToPspTime(&statp->st_ctime, creation_time_ticks);
 
     return 0;
 }
 
-int stat_file_by_fd(IOState &io, const int fd, SceIoStat *statp, const char *pref_path, uint64_t base_tick, const char *export_name) {
-    assert(statp != NULL);
+int stat_file_by_fd(IOState &io, const SceUID fd, SceIoStat *statp, const std::string &pref_path, SceUInt64 base_tick, const char *export_name) {
+    assert(statp != nullptr);
```

Comment:
I would probably write just `assert(!statp);` but if the original writer was doing this to be more clear I understand.